tests/DCPS/Thrasher/PublisherService.cpp

```diff
@@ -0,0 +1,58 @@
+
+#include "PublisherService.h"
+
+#include <string>
+
+PublisherService::PublisherService(const long domain_id, std::size_t samples_per_thread, bool durable)
+  : domain_id_(domain_id)
+  , samples_per_thread_(samples_per_thread)
+  , durable_(durable)
+  , mutex_()
+  , thread_index_(0)
+{
+  ACE_DEBUG((LM_INFO, ACE_TEXT("(%P|%t) -> PublisherService::PublisherService\n")));
+}
+
+PublisherService::~PublisherService()
+{
+  ACE_DEBUG((LM_INFO, ACE_TEXT("(%P|%t) <- PublisherService::~PublisherService\n")));
+}
+
+bool PublisherService::start(const int n_threads, const long flags)
+{
+  ACE_DEBUG((LM_INFO, ACE_TEXT("(%P|%t) -> PublisherService::start (%d threads)\n"), n_threads));
+  const int r = activate(flags, n_threads);
+  if (r != 0) {
+    ACE_DEBUG((LM_INFO, ACE_TEXT("(%P|%t) ERROR: PublisherService::start - activate() returned %d\n"), r));
+  }
+  return r == 0;
+}
+
+bool PublisherService::end()
+{
+  ACE_DEBUG((LM_INFO, ACE_TEXT("(%P|%t) <- PublisherService::end\n")));
+  const int r = wait();
+  if (r != 0) {
+    ACE_DEBUG((LM_INFO, ACE_TEXT("(%P|%t) ERROR: PublisherService::end - wait() returned %d\n"), r));
+  }
+  return r == 0;
+}
+
+int PublisherService::svc()
+{
+  try {
+    Publisher::Ptr pub(createPublisher());
+    ACE_OS::sleep(ACE_Time_Value(0, 50000)); // to simulate concurrency
```

Comment:
Concurrency can happen among multiple threads within a single process and among multiple threads among multiple processes. Since we changed the Thrasher test from a multi-process test to a single process test, there are some subtle but non-deterministic differences here.

Threads can not determine their execution order or time slots. They can make suggestions, such as changing thread priorities, or voluntarily giving up execution time. A technique to test multi-threaded application is to put sleep(0) in certain places in the code to see if it could reveal any problems, since sleep(0), though does not delay much, may give up the time slot. One can run a test many times, but what's the chance that an execution slot actually ended in a concerned place?

See the concurrency difference in test output with/without the sleep statements:

[1] without sleep: 
//...
(4129|123145373245440) pub30->started
(4129|123145373245440) pub30  12% (2 samples sent)
(4129|123145373245440) pub30  25% (4 samples sent)
(4129|123145373245440) pub30  31% (5 samples sent)
(4129|123145373245440) pub30  43% (7 samples sent)
(4129|123145373245440) pub30  50% (8 samples sent)
(4129|123145373245440) pub30  62% (10 samples sent)
(4129|123145373245440) pub30  75% (12 samples sent)
(4129|123145373245440) pub30  81% (13 samples sent)
(4129|123145373245440) pub30  93% (15 samples sent)
(4129|123145373245440) pub30  100% (16 samples sent)
(4129|123145373245440) pub30->wait_match()
(4129|123145373782016) pub31->transport cfg_31
//...

[2] with sleep:
//...
(94125|123145480273920) pub30->started
(94125|123145477591040) pub21  12% (4 samples sent)
(94125|123145467396096) pub0  40% (13 samples sent)
(94125|123145482956800) pub31->transport cfg_31
(94125|123145467932672) pub1  40% (13 samples sent)
(94125|123145475444736) pub17  21% (7 samples sent)
(94125|123145471688704) pub10  31% (10 samples sent)
(94125|123145478127616) pub22  12% (4 samples sent)
(94125|123145466859520) pub3  40% (13 samples sent)
(94125|123145466322944) pub2  40% (13 samples sent)
(94125|123145472761856) pub12  31% (10 samples sent)
(94125|123145472225280) pub11  31% (10 samples sent)
(94125|123145475981312) pub18  21% (7 samples sent)
//...

Reply:
Again, we don't care if one publisher finishes before the others.

Reply:
I think adding these sleeps (they were not in the original test) is a mistake because it misses the original purpose of the Trasher test, which was intentionally designed to have poor thread interleaving for the sake of simulating [thrashing](https://en.wikipedia.org/wiki/Thrashing_(computer_science)). Even if that goal wasn't met directly (I'm not sure this test actual achieves true thrashing), it did seem to have the effect of testing long delays in thread context switching when large amounts of data were being processed. I've made suggestions to update the readme to reflect as much.

Reply:
I understand what you are saying about 'thrashing' in your suggested changes in the readme.txt file and in your code review comments. It seems that concurrent publishing is a requirement of the Thrasher test, and the point in discussion here is how, not whether, to implement it, and also how we should accept the test results of the implementation.

There are some differences in concurrency between the real world multiple publisher processes interacting with OpenDDS and a test's multiple publisher threads. A real world process can be created randomly in terms of what other processes are doing, while the test's threads are created in a synchronized (or asynchronized) fashion with an index (unique within the process) assigned to each of them to track their individual results. The real world processes can publish continuously and therefore may have more concurrent activities of creation, publishing, or exiting, from other processes. A test tread publishes a number of messages and may have less concurrency with other threads.

Also consider the difference between a test and its unit under test (UUT), in this case, OpenDDS. Since we can not always achieve, if at all, the real world concurrency pressure in this simple test, we have some options. One is to just put the publishers in separate threads. After creation, these threads will have some concurrent activities before they exit. If they end up in, most of the time, each thread sending all its messages without interruption, we accept the results as sufficient Thrasher test to OpenDDS. 

Another approach is to use very short sleeps (currently 1/20 second and 1/100 second) in the test to encourage concurrency in order to put thrashing pressure on the UUT. If they end up in some threads still sending all its messages without interruption, we accept the results as sufficient Thrasher test to OpenDDS.

The concurrency difference in test output listed above between these two approaches does not suggest that each thread in a test run will behave exactly the same way, but they can be the mode more often than not. It's obvious which approach puts more thrashing pressure on OpenDDS as sending and receiving messages are concerned.

Again, concurrent publishing is a requirement of testing. However, since I'm out numbered in voting for the different implementations and accepting the different test results, I will make the changes.